linkerd/io/src/lib.rs

```diff
@@ -0,0 +1,58 @@
+mod boxed;
+mod peek;
+mod prefixed;
+
+pub use self::{boxed::BoxedIo, peek::Peek, prefixed::PrefixedIo};
+pub use std::io::{Error, Read, Result, Write};
+pub use tokio::io::{AsyncRead, AsyncWrite};
+
+pub type Poll<T> = futures::Poll<T, Error>;
+
+mod internal {
+    use super::{AsyncRead, AsyncWrite, Poll, Result};
+    use bytes::Buf;
+    use std::net::Shutdown;
+
+    /// This trait is private, since it's purpose is for creating a dynamic
+    /// trait object, but doing so without care can lead not getting vectored
+    /// writes.
+    ///
+    /// Instead, use the concrete `BoxedIo` type.
+    pub trait Io: AsyncRead + AsyncWrite + Send {
+        fn shutdown_write(&mut self) -> Result<()>;
+
+        /// This method is to allow using `Async::write_buf` even through a
```

Comment:
I don't understand this song and dance.  Where is the non-vectorized implementation of write_buf?  Is it the default implementation on the `AsyncWrite` trait?  why would a call to `write_buf` on a `Box<dyn AsyncWrite>` dispatch to the default implementation rather than the overriding implementation?

Reply:
> Is it the default implementation on the AsyncWrite trait?

Yep. If an impl doesn't override it, perf suffers _severely_. This isn't new code, fwiw.